tests/test_url_handler.py

```diff
@@ -0,0 +1,700 @@
+
+
+@pytest.mark.parametrize('surt, expected', [
+    ('http://(,)', {}),
+    ('http://(org,)', {'org': [(x, None) for x in alnum_list]})
+])
+def test_alphabetical_browse_domains_not_found(surt, expected):
+    project = factories.ProjectFactory()
+    factories.SURTFactory(url_project=project, value=surt)
+    returned = url_handler.alphabetical_browse(project)
+
+    assert returned == expected
+
+
+def test_alphabetical_browse_fake_project():
+    project = 'not a real project'
+    with pytest.raises(http.Http404):
+        url_handler.alphabetical_browse(project)
+
+
+def test_get_metadata():
+    project = factories.ProjectFactory()
+    vals = [x.metadata for x in factories.MetadataValuesFactory.create_batch(3)]
+    metadata = [factories.ProjectMetadataFactory(project=project, metadata=x) for x in vals]
+    expected = [(x, [models.Metadata_Values.objects.get(metadata=x).value]) for x in metadata]
+    returned = url_handler.get_metadata(project)
+
+    for each in returned:
+        assert str(each) in str(expected)
+
+
+def test_get_metadata_with_valueset():
+    project = factories.ProjectFactory()
+    valset = factories.ValuesetFactory()
+    vals = [factories.ValuesetValuesFactory(valueset=valset).value for i in range(3)]
+    metadata = factories.MetadataFactory(value_sets=[valset])
+    factories.ProjectMetadataFactory(project=project, metadata=metadata)
+    returned = url_handler.get_metadata(project)
+
+    for _, value_iter in returned:
+        for value in value_iter:
+            assert value in vals
+
+
+def test_handle_metadata(rf):
+    request = rf.post(
+        '/',
+        {
+            'color': ['blue', 'other_specify'],
+            'group': ['alpha', 'other_specify'],
+            'name': ['Dan', 'Sara'],
+            'title': ['other_specify'],
+            'job': ['other_specify'],
+        }
+    )
+    posted_data = {
+        'color': ['blue', 'other_specify'],
+        'color_other': 'magenta',
+        'group': ['alpha', 'other_specify'],
+        'name': ['Dan', 'Sara'],
+        'title': 'other_specify',
+        'title_other': 'Monsieur',
+        'job': 'other_specify',
+    }
+    expected = {
+        'color': ['blue', 'magenta'],
+        'color_other': 'magenta',
+        'group': ['alpha'],
+        'name': ['Dan', 'Sara'],
+        'title': 'Monsieur',
+        'title_other': 'Monsieur',
+        # Is this really what we want returned if the 'other' attribute isn't set?
+        'job': 'other_specify',
+    }
+
+    assert url_handler.handle_metadata(request, posted_data) == expected
+
+
+@pytest.mark.parametrize('date_str', [
+    '2006-10-25',
+    '10/25/2006',
+    '10/25/06',
+    'Oct 25 2006',
+    'Oct 25, 2006',
+    '25 Oct 2006',
+    '25 Oct, 2006',
+    'October 25 2006',
+    'October 25, 2006',
+    '25 October 2006',
+    '25 October, 2006'
+])
+def test_validate_date_with_valid_dates(date_str):
+    assert url_handler.validate_date(date_str) is not None
```

Comment:
We could test the result is an instance of datetime.date rather than just verifying it is not None.